pkgs/applications/virtualization/qemu/encode-video.c

```diff
@@ -0,0 +1,737 @@
+{
+    void *buf;
+    int i, readlen;
+
+    is_error = false;
+
+    if ((buf = malloc(len)) == NULL) {
+        fprintf(stderr, "Unable to allocate buffer for "
+                "intermediate video packet: %s\n", strerror(errno));
+        is_error = true;
+        return NULL;
+    }
+
+    for (i = 0; i < len; i += readlen) {
+        readlen = gzread(ifile, buf + i, len - i);
+        if (readlen <= 0) {
+            free(buf);
+            /* Don't print an error because we want to make sure that whenever
+             * the stream ends prematurely, we still do have a valid video.
+             */
+            if (readlen < 0) {
+                fprintf(stderr, "Unable to read %zu bytes from input file.",
+                        len - i);
+                is_error = true;
+            }
+            return NULL;
+        }
+    }
+
+    return buf;
+}
+
+/* The packed pixel format from QEMU (and thus Pixman) is using 2 or 4 byte
+ * sequences which are in native host byte order. FFmpeg on the other hand
+ * expects 8-bit sequences, so we need to convert it to big endian as all of
+ * the pixel formats we've choosen for FFmpeg are big endian (either explicitly
+ * or implicitly).
+ */
+static bool convert_endian(void *data, size_t len, uint8_t bpp)
+{
+    const uint16_t endian_test = 1;
+
+    if (*(uint8_t*)&endian_test == 0)
+        return true;
+
+    switch (bpp) {
+        case 2:
+            for (int i = 0; i < len; ++i, data += bpp)
+                *(uint16_t*)data = bswap_16(*(uint16_t*)data);
+            break;
+        case 4:
+            for (int i = 0; i < len; ++i, data += bpp)
+                *(uint32_t*)data = bswap_32(*(uint32_t*)data);
+            break;
+        default:
+            fprintf(stderr, "Unable to handle pixel byte size of %d.\n", bpp);
+            return false;
+    }
+    return true;
+}
+
+static AVFrame *alloc_frame(enum AVPixelFormat pix_fmt, uint32_t width,
+                            uint32_t height)
+{
+    AVFrame *frame;
+
+    if ((frame = av_frame_alloc()) == NULL)
+        return NULL;
+
+    frame->format = pix_fmt;
+    frame->width  = width;
+    frame->height = height;
+
+    if (av_frame_get_buffer(frame, 0) < 0) {
+        fprintf(stderr, "Could not allocate frame data.\n");
+        return NULL;
+    }
+
+    return frame;
+}
+
+/* Parse a switch packet (indicated by an 'S' byte).
+ *
+ * The format is (numbers after colons are bit sizes):
+ *
+ * <<Width:32, Height:32, Format:8, BPP:8>>
+ */
+static struct packet_switch *parse_switch(gzFile ifile)
+{
+    void *buf;
+    struct packet_switch *out;
+    uint8_t tmp_format;
+
+    if ((buf = alloc_read(ifile, 10)) == NULL)
+        return NULL;
+
+    if ((out = malloc(sizeof(struct packet_switch))) == NULL) {
+        fprintf(stderr, "Unable to allocate packet_switch: %s\n",
+                strerror(errno));
+        is_error = true;
+        free(buf);
+        return NULL;
+    }
+
+    memcpy(&out->width,  buf,     4);
```

Comment:
some `sizeof` instead of raw numbers would be nice to see here

Reply:
You're right, I thought this was apparent due to the Erlang bitstring above.